onnxscript/ir/_core.py

```diff
@@ -1977,8 +1978,63 @@ def insert_before(self, node: Node, new_nodes: Iterable[Node] | Node, /) -> None
         self._nodes.insert_before(node, new_nodes)
 
     def sort(self) -> None:
-        """Topologically sort the nodes in the graph."""
-        raise NotImplementedError("Not implemented yet")
+        """Perform a topological sort of this graph and all its subgraphs in O(#nodes + #values) time.
+
+        This method sorts the nodes that each node appears AFTER all nodes it depend on, while maintaining
+        the original order as much as possible.
+
+        Raises:
+            ValueError: If the graph contains a cycle, making topological sorting impossible.
+        """
+        # Perform a topological sort of the graph using Kahn's algorithm plus a priority queue.
+        # 1. Initialize: Track each node's graph, depth (number of direct children), direct parent nodes, and original index.
+        # 2. Priority Queue: Track nodes with zero children in a priority queue, using their original index for ordering. This
+        # ensures that nodes appearing LATER in the original order are processed EARLIER (the order will be reversed at the end).
+        # 3. Topological Sort:
+        #   - While the priority queue is not empty:
+        #     - Pop the node with the largest index and add it to the sorted nodes by graph.
+        #     - Decrement the depth of its parent nodes. If any parent node has zero direct children, push it into the queue.
+        # 4. Cycle Check: Ensure all nodes are processed. If not, raise a ValueError indicating a cycle.
+        # 5. Finishing: Reverse the sorted nodes in each subgraph to get the topological order.
+        
+        # Initialize
+        nodes = list(onnxscript.ir.traversal.RecursiveGraphIterator(self))
+        sorted_nodes_by_graph: dict[Graph, list[Node]] = {graph: [] for graph in {node.graph for node in nodes}}
```

Comment:
Just checking. Was the implicit dependencies from subgraphs handled? This will affect the topological ordering.